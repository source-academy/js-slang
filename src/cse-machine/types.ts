import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
