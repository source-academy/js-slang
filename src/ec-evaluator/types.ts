import * as es from 'estree'

import { Context } from '..'
import { Environment } from '../types'
import { Agenda, Stash } from './interpreter'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
  srcNode: es.Node
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
  srcNode: es.Node
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
  srcNode: es.Node
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
  srcNode: es.Node
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

export type CmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => void

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
