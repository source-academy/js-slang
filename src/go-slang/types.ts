export enum NodeType {
  ExpressionStatement = 'ExpressionStatement',
  Literal = 'Literal',
  UnaryExpression = 'UnaryExpression',
  BinaryExpression = 'BinaryExpression'
}

// NOTE: right now, the only type of source file is an expression statement
export type SourceFile = ExpressionStatement

type Expression = Literal | UnaryExpression | BinaryExpression

export interface Node {
  type: NodeType
}

export interface ExpressionStatement extends Node {
  type: NodeType.ExpressionStatement
  expression: Expression
}

export interface Literal extends Node {
  type: NodeType.Literal
  value: any
}

export type UnaryOperator = '+' | '-'

export interface UnaryExpression extends Node {
  type: NodeType.UnaryExpression
  operator: UnaryOperator
  argument: Expression
}

export type BinaryOperator =
  | '+'
  | '-'
  | '*'
  | '/'
  | '%'
  | '|'
  | '^'
  | '=='
  | '!='
  | '<'
  | '<='
  | '>'
  | '>='

export interface BinaryExpression extends Node {
  type: NodeType.BinaryExpression
  operator: BinaryOperator
  left: Expression
  right: Expression
}

export enum CommandType {
  UnaryOp = 'UnaryOp',
  BinaryOp = 'BinaryOp'
}

export interface Command {
  type: CommandType
}

export interface UnaryOp extends Command {
  type: CommandType.UnaryOp
  operator: UnaryOperator
}

export interface BinaryOp extends Command {
  type: CommandType.BinaryOp
  operator: BinaryOperator
}

export type Instruction = ExpressionStatement | Expression | UnaryOp | BinaryOp
