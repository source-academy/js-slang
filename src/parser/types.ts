import type { Program } from 'estree'

import type { Context, Chapter, Node, SourceError, Variant } from '../types'

export type { Options as AcornOptions } from 'acorn'
export type { ParserOptions as BabelOptions } from '@babel/parser'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}

export interface Rule<T extends Node> {
  name: string
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
