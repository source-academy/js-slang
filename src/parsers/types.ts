import { ParserOptions } from '@babel/parser'
import { Options } from 'acorn'
import { Program } from 'estree'

import { Context } from '../types'

export type AcornOptions = Options
export type BabelOptions = ParserOptions

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
