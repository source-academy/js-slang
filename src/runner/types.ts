import type { Program } from 'estree'
import type { Context, IOptions, Result } from '..'

export type Runner = (program: Program, context: Context, options: IOptions) => Promise<Result>
