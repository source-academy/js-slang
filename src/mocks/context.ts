import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
