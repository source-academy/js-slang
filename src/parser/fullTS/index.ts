import { parse as babelParse } from '@babel/parser'
import { createProjectSync, ts } from '@ts-morph/bootstrap'
import { Program } from 'estree'

import { Context } from '../..'
import * as TypedES from '../../typeChecker/tsESTree'
import { removeTSNodes } from '../../typeChecker/typeErrorChecker'
import { FatalSyntaxError } from '../errors'
import { transformBabelASTToESTreeCompliantAST } from '../source/typed/utils'
import { AcornOptions, Parser } from '../types'
import { defaultBabelOptions, positionToSourceLocation } from '../utils'

const IMPORT_TOP_LEVEL_ERROR =
  'An import declaration can only be used at the top level of a namespace or module.'
const START_OF_MODULE_ERROR = 'Cannot find module '

export class FullTSParser implements Parser<AcornOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<AcornOptions>,
    throwOnError?: boolean
  ): Program | null {
    let code = ''
    // Add builtins to code
    // Each declaration is replaced with a single constant declaration with type `any`
    // to reduce evaluation time
    for (const builtin of context.nativeStorage.builtins) {
      code += `const ${builtin[0]}: any = 1\n`
    }
    // Add prelude functions to code
    // Each declaration is replaced with a single constant declaration with type `any`
    // to reduce evaluation time
    if (context.prelude) {
      const preludeFns = context.prelude.split('\nfunction ').slice(1)
      preludeFns.forEach(fnString => {
        const fnName = fnString.split('(')[0]
        // Functions in prelude that start with $ are not added
        if (fnName.startsWith('$')) {
          return
        }
        code += `const ${fnName}: any = 1\n`
      })
    }
    // Get line offset
    const lineOffset = code.split('\n').length - 1

    // Add program string to code string,
    // wrapping it in a block to allow redeclaration of variables
    code = code + '{' + programStr + '}'
    // Initialize file to analyze
    const project = createProjectSync({ useInMemoryFileSystem: true })
    const filename = 'program.ts'
    project.createSourceFile(filename, code)

    // Get TS diagnostics from file, formatted as TS error string
    const diagnostics = ts.getPreEmitDiagnostics(project.createProgram())
    const formattedString = project.formatDiagnosticsWithColorAndContext(diagnostics)

    // Reformat TS error string to Source error by getting line number using regex
    // This is because logic to retrieve line number is only present in
    // formatDiagnosticsWithColorAndContext and cannot be called directly
    const lineNumRegex = /(?<=\[7m)\d+/
    diagnostics.forEach(diagnostic => {
      const message = diagnostic.messageText.toString()
      // Ignore errors regarding imports
      // as TS does not have information about Source modules
      if (message === IMPORT_TOP_LEVEL_ERROR || message.startsWith(START_OF_MODULE_ERROR)) {
        return
      }
      const lineNumRegExpArr = lineNumRegex.exec(formattedString.split(message)[1])
      const lineNum = (lineNumRegExpArr === null ? 0 : parseInt(lineNumRegExpArr[0])) - lineOffset
      // Ignore any errors that occur in builtins/prelude (line number <= 0)
      if (lineNum <= 0) {
        return
      }
      const position = { line: lineNum, column: 0, offset: 0 }
      context.errors.push(new FatalSyntaxError(positionToSourceLocation(position), message))
    })

    if (context.errors.length > 0) {
      return null
    }

    // Parse code into Babel AST, which supports type syntax
    const ast = babelParse(programStr, {
      ...defaultBabelOptions,
      sourceFilename: options?.sourceFile,
      errorRecovery: throwOnError ?? true
    })

    if (ast.errors.length) {
      ast.errors
        .filter(error => error instanceof SyntaxError)
        .forEach(error => {
          context.errors.push(
            new FatalSyntaxError(
              positionToSourceLocation((error as any).loc, options?.sourceFile),
              error.toString()
            )
          )
        })

      return null
    }

    // Transform Babel AST into ESTree AST
    const typedProgram: TypedES.Program = ast.program as TypedES.Program
    const transpiledProgram: Program = removeTSNodes(typedProgram)
    transformBabelASTToESTreeCompliantAST(transpiledProgram)

    return transpiledProgram
  }

  validate(_ast: Program, _context: Context, _throwOnError: boolean): boolean {
    return true
  }

  toString(): string {
    return 'FullTSParser'
  }
}
