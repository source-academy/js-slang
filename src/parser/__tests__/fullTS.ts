import { parseError } from '../..'
import { mockContext } from '../../mocks/context'
import { Chapter } from '../../types'
import { FullTSParser } from '../fullTS'

const parser = new FullTSParser()
let context = mockContext(Chapter.FULL_TS)

beforeEach(() => {
  context = mockContext(Chapter.FULL_TS)
})

describe('fullTS parser', () => {
  it('formats errors correctly', () => {
    const code = `type StringOrNumber = string | number;
      const x: StringOrNumber = true;
    `

    parser.parse(code, context)
    expect(parseError(context.errors)).toMatchInlineSnapshot(
      `"Line 2: Type \'boolean\' is not assignable to type \'StringOrNumber\'."`
    )
  })

  it('allows usage of builtins/preludes', () => {
    const code = `const xs = list(1);
      const ys = list(1);
      equal(xs, ys);
    `

    parser.parse(code, context)
    expect(parseError(context.errors)).toMatchInlineSnapshot(`""`)
  })

  it('allows usage of imports/modules', () => {
    const code = `import { show, heart } from "rune";
      show(heart);
    `

    parser.parse(code, context)
    expect(parseError(context.errors)).toMatchInlineSnapshot(`""`)
  })

  it('returns ESTree compliant program', () => {
    const code = `type StringOrNumber = string | number;
      const x: StringOrNumber = 1;
    `

    // Resulting program should not have node for type alias declaration
    const parsedProgram = parser.parse(code, context)
    expect(parsedProgram).toMatchSnapshot()
  })
})
