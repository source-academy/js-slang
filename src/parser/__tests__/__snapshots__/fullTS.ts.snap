// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fullTS parser returns ESTree compliant program 1`] = `
Node {
  "body": Array [
    Node {
      "declarations": Array [
        Node {
          "end": 72,
          "id": Node {
            "end": 68,
            "loc": SourceLocation {
              "end": Position {
                "column": 29,
                "line": 2,
              },
              "identifierName": "x",
              "source": undefined,
              "start": Position {
                "column": 12,
                "line": 2,
              },
            },
            "name": "x",
            "start": 51,
            "type": "Identifier",
            "typeAnnotation": Node {
              "end": 68,
              "loc": SourceLocation {
                "end": Position {
                  "column": 29,
                  "line": 2,
                },
                "identifierName": undefined,
                "source": undefined,
                "start": Position {
                  "column": 13,
                  "line": 2,
                },
              },
              "start": 52,
              "type": "TSTypeAnnotation",
              "typeAnnotation": Node {
                "end": 68,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 29,
                    "line": 2,
                  },
                  "identifierName": undefined,
                  "source": undefined,
                  "start": Position {
                    "column": 15,
                    "line": 2,
                  },
                },
                "start": 54,
                "type": "TSTypeReference",
                "typeName": Node {
                  "end": 68,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 29,
                      "line": 2,
                    },
                    "identifierName": "StringOrNumber",
                    "source": undefined,
                    "start": Position {
                      "column": 15,
                      "line": 2,
                    },
                  },
                  "name": "StringOrNumber",
                  "start": 54,
                  "type": "Identifier",
                },
              },
            },
          },
          "init": Node {
            "end": 72,
            "loc": SourceLocation {
              "end": Position {
                "column": 33,
                "line": 2,
              },
              "identifierName": undefined,
              "source": undefined,
              "start": Position {
                "column": 32,
                "line": 2,
              },
            },
            "raw": "1",
            "start": 71,
            "type": "Literal",
            "value": 1,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 33,
              "line": 2,
            },
            "identifierName": undefined,
            "source": undefined,
            "start": Position {
              "column": 12,
              "line": 2,
            },
          },
          "start": 51,
          "type": "VariableDeclarator",
        },
      ],
      "end": 73,
      "kind": "const",
      "loc": SourceLocation {
        "end": Position {
          "column": 34,
          "line": 2,
        },
        "identifierName": undefined,
        "source": undefined,
        "start": Position {
          "column": 6,
          "line": 2,
        },
      },
      "start": 45,
      "type": "VariableDeclaration",
    },
  ],
  "end": 78,
  "interpreter": null,
  "loc": SourceLocation {
    "end": Position {
      "column": 4,
      "line": 3,
    },
    "identifierName": undefined,
    "source": undefined,
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "sourceType": "module",
  "start": 0,
  "type": "Program",
}
`;
